currying/README.md

```diff
@@ -0,0 +1,183 @@
+    Book shining = kingHorrorBooksFunc.withTitle("The Shining")
+    .withPublicationDate(LocalDate.of(1977, 1, 28));
+    Book darkTower = kingFantasyBooksFunc.withTitle("The Dark Tower: Gunslinger")
+    .withPublicationDate(LocalDate.of(1982, 6, 10));
+
+    // Creates fantasy books by J.K. Rowling
+    Book chamberOfSecrets = rowlingFantasyBooksFunc.withTitle("Harry Potter and the Chamber of Secrets")
+    .withPublicationDate(LocalDate.of(1998, 7, 2));
+
+    // Create sci-fi books
+    Book dune = scifiBookFunc.withAuthor("Frank Herbert")
+    .withTitle("Dune")
+    .withPublicationDate(LocalDate.of(1965, 8, 1));
+    Book foundation = scifiBookFunc.withAuthor("Isaac Asimov")
+    .withTitle("Foundation")
+    .withPublicationDate(LocalDate.of(1942, 5, 1));
+
+    LOGGER.info("Stephen King Books:");
+    LOGGER.info(shining.toString());
+    LOGGER.info(darkTower.toString());
+
+    LOGGER.info("J.K. Rowling Books:");
+    LOGGER.info(chamberOfSecrets.toString());
+
+    LOGGER.info("Sci-fi Books:");
+    LOGGER.info(dune.toString());
+    LOGGER.info(foundation.toString());
+}
+```
+Program output:
+```
+Librarian begins their work.
+Stephen King Books:
+Book{genre=HORROR, author='Stephen King', title='The Shining', publicationDate=1977-01-28}
+Book{genre=FANTASY, author='Stephen King', title='The Dark Tower: Gunslinger', publicationDate=1982-06-10}
+J.K. Rowling Books:
+Book{genre=FANTASY, author='J.K. Rowling', title='Harry Potter and the Chamber of Secrets', publicationDate=1998-07-02}
+Sci-fi Books:
+Book{genre=SCI_FI, author='Frank Herbert', title='Dune', publicationDate=1965-08-01}
+Book{genre=SCI_FI, author='Isaac Asimov', title='Foundation', publicationDate=1942-05-01}
+```
+
+## Class diagram
+![currying-uml](./etc/currying.urm.png)
+
+## Applicability
+A curried function which has only been passed some of its arguments is called a partial application. Partial application 
+allows for the creation of functions with some pre-defined data in their scope, since partial application can be used to
+create specialised functions with lower arity. This abstraction can help keep code readable and concise. Therefore, currying is useful when frequently calling functions with fixed parameters.
+
+## Known uses
+Most functional programming languages support curried functions. A popular example is [Haskell](https://www.haskell.org/), in which all functions are considered curried. 
+
+## Consequences
+Pros
+* Currying allows for partial application, which can be used to create specialised functions concisely.
+
+Cons
+* The order of the parameters in a curried function is important since we want to take advantage of partial application. It is best to input the most general parameters first and input specific parameters last. 
+* As shown in the programmatic example above, curried functions with several parameters have a cumbersome type signature (in Java).
+
+## Credits
+* [Currying in Java](https://www.baeldung.com/java-currying)
```

Comment:
For related patterns, perhaps the builder pattern and functional interfaces patterns you mentioned within your write up?

Reply:
This is a good point. I will add the builder pattern. However, I don't think functional interfaces are a pattern. So I will not include them.